### PR TITLE
Added File Churn Metric

### PR DESCRIPTION
**Pull request by o2sh**
This pull request, part of #1059, introduces a new info line to onefetch called "Churn." This info line displays the files with the most modifications (commits), providing valuable insights into code volatility and potential hotspots.

Calculating the churn is computationally expensive 😢 , as it requires comparing each commit's tree with its parent tree to obtain the diff and see which files were modified. Since git (and gitoxide) does not store deltas, this process becomes resource-intensive the more commits there are in the git history.

To optimize performance, I added a limit on the number of commits used to compute the file churns. This limit can be configured via a CLI flag:

```
--churn-commit-limit <NUM>     NUM of commits from HEAD used to compute file churns [default: 100]
``` 

Here is what it looks like on onefetch:

image

### CODE REVIEW

**Comment by Byron** [comment]
Thanks for reeling me in :)!

I like the implementation even though I'd love it if there was a way to 'more smoothly' deal with diffs against the first tree, which is the empty tree.

Besides that, I hope you excuse me pushing improvements directly into this branch as I think it's easier to see what I mean that way, with the disclaimer of: these commits are suggestions, please feel free to change or remove as you see fit no questions asked.

With all that out of the way, the commit here already improves performance quite a bit:

```
linux ( master)
❯ hyperfine --warmup 1 /Users/byron/dev/github.com/o2sh/onefetch/onefetch /Users/byron/dev/github.com/o2sh/onefetch/onefetch-optimized
Benchmark 1: /Users/byron/dev/github.com/o2sh/onefetch/onefetch
  Time (mean ± σ):     20.285 s ±  0.455 s    [User: 22.086 s, System: 2.658 s]
  Range (min … max):   19.617 s … 21.018 s    10 runs

Benchmark 2: /Users/byron/dev/github.com/o2sh/onefetch/onefetch-optimized
  Time (mean ± σ):     14.635 s ±  0.311 s    [User: 16.232 s, System: 2.659 s]
  Range (min … max):   13.949 s … 15.044 s    10 runs

Summary
  '/Users/byron/dev/github.com/o2sh/onefetch/onefetch-optimized' ran
    1.39 ± 0.04 times faster than '/Users/byron/dev/github.com/o2sh/onefetch/onefetch'
```

---- 

That's not all though, as I am working on a faster form of traversal which can make use of commitgraph data-structures - stay tuned (a new `gix` release is necessary) :).

**Review by Byron — APPROVED** [review]
I had to dial back my ambition a little as we are using that commit author for every commit, which means we can't leverage the commitgraph anymore.

However, I found another opportunity which leads to another 5ish % of performance improvement.

It looks good to me now!

```
linux ( master) +369 -819 [!?] took 5m27s
❯ hyperfine --warmup 1 /Users/byron/dev/github.com/o2sh/onefetch/onefetch-optimized /Users/byron/dev/github.com/o2sh/onefetch/onefetch-optimized-v2
Benchmark 1: /Users/byron/dev/github.com/o2sh/onefetch/onefetch-optimized
  Time (mean ± σ):     14.540 s ±  0.189 s    [User: 15.940 s, System: 2.735 s]
  Range (min … max):   14.307 s … 14.795 s    10 runs

Benchmark 2: /Users/byron/dev/github.com/o2sh/onefetch/onefetch-optimized-v2
  Time (mean ± σ):     13.803 s ±  0.157 s    [User: 15.227 s, System: 2.716 s]
  Range (min … max):   13.488 s … 13.969 s    10 runs

Summary
  '/Users/byron/dev/github.com/o2sh/onefetch/onefetch-optimized-v2' ran
    1.05 ± 0.02 times faster than '/Users/byron/dev/github.com/o2sh/onefetch/onefetch-optimized'
```

### Failures

Not everything worked, here is what didn't work.

<details><summary>commits</summary>

```patch
commit 471f88fc9247e374fbfdf10a6f64476214b64bde
Author: Sebastian Thiel <sebastian.thiel@icloud.com>
Date:   Sun Jun 4 16:57:41 2023 +0200

    Even more performance by avoiding allocations
    
    Previously, each commit caused us to allocate two strings, even though we drop most of them
    right after when inserting them into a map.
    
    Now we keep keep the owned copies of unique signatures only once.
    
    Warning: Since we use a hash, there is the chance of a hash collision, which would lead
             to the commit count being slightly off. I think that's acceptable.

diff --git a/src/info/utils/git.rs b/src/info/utils/git.rs
index b7a279a..b32cdeb 100644
--- a/src/info/utils/git.rs
+++ b/src/info/utils/git.rs
@@ -11,6 +11,7 @@ use gix::prelude::ObjectIdExt;
 use gix::Commit;
 use regex::Regex;
 use std::collections::HashMap;
+use std::hash::Hash;
 use std::str::FromStr;
 
 pub struct CommitMetrics {
@@ -30,9 +31,29 @@ pub struct Sig {
     email: gix::bstr::BString,
 }
 
-impl From<gix::actor::Signature> for Sig {
-    fn from(gix::actor::Signature { name, email, .. }: gix::actor::Signature) -> Self {
-        Self { name, email }
+impl Sig {
+    #[inline]
+    fn hash(sig: &gix::mailmap::snapshot::Signature<'_>) -> SigHash {
+        let mut hasher = std::collections::hash_map::DefaultHasher::new();
+        use std::hash::Hasher;
+        sig.name.hash(&mut hasher);
+        sig.email.hash(&mut hasher);
+        hasher.finish()
+    }
+}
+
+type SigHash = u64;
+
+impl From<gix::mailmap::snapshot::Signature<'_>> for Sig {
+    fn from(
+        gix::mailmap::snapshot::Signature { name, email, .. }: gix::mailmap::snapshot::Signature<
+            '_,
+        >,
+    ) -> Self {
+        Self {
+            name: name.into_owned(),
+            email: email.into_owned(),
+        }
     }
 }
 
@@ -45,7 +66,8 @@ impl CommitMetrics {
         let mut commit_iter_peekable = commit_iter.by_ref().peekable();
 
         let mailmap_config = repo.open_mailmap();
-        let mut number_of_commits_by_signature: HashMap<Sig, usize> = HashMap::new();
+        let mut signature_map: HashMap<SigHash, Sig> = HashMap::new();
+        let mut number_of_commits_by_signature: HashMap<SigHash, usize> = HashMap::new();
         let mut total_number_of_commits = 0;
         let mut number_of_commits_by_file_path: HashMap<BString, usize> = HashMap::new();
 
@@ -57,13 +79,17 @@ impl CommitMetrics {
                 continue;
             }
 
-            let sig = Sig::from(mailmap_config.resolve(commit.author()?));
+            let sig = mailmap_config.resolve_cow(commit.author()?);
+            let sig_hash = Sig::hash(&sig);
+            let sig = signature_map
+                .entry(sig_hash)
+                .or_insert_with(|| Sig::from(sig));
 
             if is_bot(&sig.name, &bot_regex_pattern) {
                 continue;
             }
 
-            *number_of_commits_by_signature.entry(sig).or_insert(0) += 1;
+            *number_of_commits_by_signature.entry(sig_hash).or_insert(0) += 1;
 
             if total_number_of_commits <= options.info.churn_commit_limit {
                 compute_diff(&mut number_of_commits_by_file_path, &commit, repo)?;
@@ -83,6 +109,7 @@ impl CommitMetrics {
         }
 
         let (authors_to_display, total_number_of_authors) = compute_authors(
+            signature_map,
             number_of_commits_by_signature,
             total_number_of_commits,
             options.info.number_of_authors,
@@ -133,7 +160,8 @@ fn compute_file_churns(
 }
 
 fn compute_authors(
-    number_of_commits_by_signature: HashMap<Sig, usize>,
+    mut signature_map: HashMap<SigHash, Sig>,
+    number_of_commits_by_signature: HashMap<SigHash, usize>,
     total_number_of_commits: usize,
     number_of_authors_to_display: usize,
     show_email: bool,
@@ -141,7 +169,17 @@ fn compute_authors(
 ) -> (Vec<Author>, usize) {
     let total_number_of_authors = number_of_commits_by_signature.len();
     let mut signature_with_number_of_commits_sorted: Vec<(Sig, usize)> =
-        number_of_commits_by_signature.into_iter().collect();
+        number_of_commits_by_signature
+            .into_iter()
+            .map(|(sig_hash, count)| {
+                (
+                    signature_map
+                        .remove(&sig_hash)
+                        .expect("each signature is available in the map"),
+                    count,
+                )
+            })
+            .collect();
 
     signature_with_number_of_commits_sorted.sort_by(|(sa, a_count), (sb, b_count)| {
         b_count.cmp(a_count).then_with(|| sa.name.cmp(&sb.name))
@@ -235,6 +273,7 @@ fn is_bot(author_name: &BString, bot_regex_pattern: &Option<MyRegex>) -> bool {
 #[cfg(test)]
 mod tests {
     use super::*;
+    use std::borrow::Cow;
 
     #[test]
     fn test_get_no_bots_regex() {
@@ -257,34 +296,47 @@ mod tests {
 
     #[test]
     fn test_compute_authors() {
-        let mut number_of_commits_by_signature: HashMap<Sig, usize> = HashMap::new();
-        number_of_commits_by_signature.insert(
-            Sig {
-                name: "John Doe".into(),
-                email: "johndoe@example.com".into(),
-            },
-            10,
-        );
-        number_of_commits_by_signature.insert(
-            Sig {
-                name: "Jane Doe".into(),
-                email: "janedoe@example.com".into(),
-            },
-            5,
-        );
-        number_of_commits_by_signature.insert(
-            Sig {
-                name: "Ellen Smith".into(),
-                email: "ellensmith@example.com".into(),
-            },
-            50,
-        );
+        let mut signature_map: HashMap<SigHash, Sig> = HashMap::new();
+        let mut number_of_commits_by_signature: HashMap<SigHash, usize> = HashMap::new();
+        for (sig, count) in [
+            (
+                Sig {
+                    name: "John Doe".into(),
+                    email: "johndoe@example.com".into(),
+                },
+                10,
+            ),
+            (
+                Sig {
+                    name: "Jane Doe".into(),
+                    email: "janedoe@example.com".into(),
+                },
+                5,
+            ),
+            (
+                Sig {
+                    name: "Ellen Smith".into(),
+                    email: "ellensmith@example.com".into(),
+                },
+                50,
+            ),
+        ] {
+            let sig_hash = Sig::hash(&gix::mailmap::snapshot::Signature {
+                name: Cow::Borrowed(sig.name.as_ref()),
+                email: Cow::Borrowed(sig.email.as_ref()),
+                time: Default::default(),
+            });
+
+            number_of_commits_by_signature.insert(sig_hash, count);
+            signature_map.insert(sig_hash, sig);
+        }
         let total_number_of_commits = 15;
         let number_of_authors_to_display = 2;
         let show_email = false;
         let number_separator = NumberSeparator::Comma;
 
         let (authors, total_number_of_authors) = compute_authors(
+            signature_map,
             number_of_commits_by_signature,
             total_number_of_commits,
             number_of_authors_to_display,
```

</details>

File: src/info/churn.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        assert_eq!(
+            truncate_file_path("path/to/file.txt", 3),
+            "path/to/file.txt"
+        );
+        assert_eq!(
+            truncate_file_path("another/file.txt", 2),
+            "another/file.txt"
+        );
+        assert_eq!(truncate_file_path("file.txt", 1), "file.txt");
+        assert_eq!(truncate_file_path("path/to/file.txt", 2), ".../to/file.txt");
```
Comment (Byron) [inline src/info/churn.rs:151]: I wonder if an actual ellipsis could be used here, as it's smaller usually which I prefer visually: `...` `…`

Comment (o2sh) [reply src/info/churn.rs]: Great suggestion, done.

**Review by spenserblack — COMMENTED** [review]
> Since git ... does not store delta

I always thought that that was exactly how git worked :laughing:

Mostly LGTM, but you might want to check out my concern about OS compatibility.

File: src/info/churn.rs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        let pad = self.title().len() + 2;
+
+        for (i, churn) in self.file_churns.iter().enumerate() {
+            let churn_str = churn.color(self.info_color);
+
+            if i == 0 {
+                let _ = write!(churn_info, "{churn_str}");
+            } else {
+                let _ = write!(churn_info, "\n{:<width$}{}", "", churn_str, width = pad);
+            }
```
[the remark this answers is not in the document]
Comment (o2sh) [reply src/info/churn.rs]: Good point! thanks

File: src/info/churn.rs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        "Churn".into()
+    }
+
+    fn should_color(&self) -> bool {
+        false
+    }
+}
+
+fn truncate_file_path(path: &str, depth: usize) -> String {
+    let components: Vec<&str> = path.split('/').collect();
```
[the remark this answers is not in the document]
Comment (o2sh) [reply src/info/churn.rs:92]: @Byron can maybe confirm, but It seems like Git stores file paths using the UNIX format and uses forward slashes ("/") as the directory separator in file paths, even on Windows systems.

**Comment by o2sh** [comment]
BTW, @spenserblack, what do you think about this new info line? Do you believe it provides value for users? I apologize for not discussing it beforehand :pensive:

Although the performance impact has been significantly reduced, thanks to @Byron, it may still be noticeable.

File: src/info/churn.rs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+    if depth == 0 || components.len() <= depth {
+        return path.to_string();
+    }
+
+    let truncated_components: Vec<&str> = components
```
[the remark this answers is not in the document]
Comment (o2sh) [reply src/info/churn.rs:98]: I'm sorry but I'm not sure I see where I can use this 😓 We're only manipulating strings in this method.

File: src/info/churn.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (spenserblack) [reply src/info/churn.rs:92]: Well that makes things easier! Learned another new thing about Git 🙂

File: src/info/churn.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (spenserblack) [reply src/info/churn.rs:98]: Yeah, maybe it's overkill. I was thinking of a way to be safe with varying OS separators, and this is mostly coming from `os.path.split` in Python. I was figuring the Rust equivalent would be something like
```rust
let path = Path::new("a\\b\\c.txt");
let components: Vec<_> = path.iter().collect();
```

**Comment by spenserblack** [comment]
If we're concerned about performance, we could consider this an experimental feature that defaults to being disabled 🙂 

But this is definitely a cool idea! :+1:

File: src/info/churn.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Byron) [reply src/info/churn.rs:92]: Yes, indeed, these `&BStr` `gix` passes around aren't `&std::path::Path` for the reason that `git` normalizes paths and stores them with `/` as component separator.

**Comment by Byron** [comment]
From a performance perspective, I think it will be fine for 99.5% of the cases. After all, it's *just* 100 commits to compute deltas for, and thanks to the performance improvements `onefetch` now has 'won' some time credits to put towards a feature like this.

However, there are repositories with huge trees, and those inevitably take longer to create diffs for as the tree-traversal has to touch so many objects. I have seen speeds of as low as 4 deltas/s per core, which means 100 commit-diffs could easily take 25s or so. It would be great if this case could automatically be handled, but even that would cost time that now always adds to the overall time. The idea was to load the index and apply some heuristic to understand how long diffs would probably take, but it seems tricky to get right. *Another* option would be to run diffs on another thread, send it only the first X commits as before, but cancel cancel it as soon as all commits have been iterated. That way, one wouldn't use any additional wallclock time to compute diffs and could abort long-running diffs pretty swiftly - the worst case diffs I have seen take 250ms, so at worst the operation would take 250ms longer as the thread will only respond after a diff was done.

Let me give that a try as this would allow the feature to be toggled on unconditionally without measurable impact.

**Comment by Byron** [comment]
I'd think this was a success :D, another ~10% faster now.

```
linux ( master) +369 -819 [!?]
❯ hyperfine --warmup 1 /Users/byron/dev/github.com/o2sh/onefetch/onefetch /Users/byron/dev/github.com/o2sh/onefetch/onefetch-optimized
Benchmark 1: /Users/byron/dev/github.com/o2sh/onefetch/onefetch
  Time (mean ± σ):     13.395 s ±  0.518 s    [User: 15.469 s, System: 2.551 s]
  Range (min … max):   12.952 s … 14.707 s    10 runs

Benchmark 2: /Users/byron/dev/github.com/o2sh/onefetch/onefetch-optimized
  Time (mean ± σ):     12.083 s ±  0.146 s    [User: 16.981 s, System: 2.482 s]
  Range (min … max):   11.833 s … 12.374 s    10 runs

Summary
  '/Users/byron/dev/github.com/o2sh/onefetch/onefetch-optimized' ran
    1.11 ± 0.04 times faster than '/Users/byron/dev/github.com/o2sh/onefetch/onefetch'
```

In the commit message, there are some ideas for further improvements related to how the feature is presented - I think it *could* be polished to show the amount of commits actually used to create the churn summary in case these were less than expected as there might not have been enough time.

**Comment by Byron** [comment]
The recent test-failure has shown that it's probably a good idea to indicate how many commits were used for churn rates if these are below the desired value, as very small repos might also trigger this case (like happened in the test). That failing test at least caused the mitigation that the thread will always produce at least one diff.

File: src/info/churn.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (o2sh) [reply src/info/churn.rs:100]: https://github.com/o2sh/onefetch/pull/1071#discussion_r1218594316

**Comment by o2sh** [comment]
Thanks to @Byron optimizations the impact on wall clock time has been reduced to nearly zero. As you suggested @Byron, I added the amount of commits actually used to create the churn summary:

image

For a more deterministic approach, users can specify the churn-pool-size:

image

image

And thanks a for the review @spenserblack

**Review by Byron — COMMENTED** [review]
I love the idea of being non-deterministic by default using the available time for churn-computation, while optionally offering a desired amount that should be met if set.

However, this comes at great costs: Running `onefetch` with these semantics on the linux kernel easily doubles its real-memory footprint.

Screenshot 2023-06-09 at 07 36 35

The reason for this is that we have no bound for how many detached objects are stuffed into the channel, which was the case previously and only that's why I deemed it correct to send the whole object buffer.

With these semantics, one has to reduce the costs of the commits sent, by re-retrieving the commit in the thread. This itself comes at a cost though, even though it is probably negligible.

I have implemented the alternative that sends IDs instead of data, which does bring down overall memory usage (while definitely still being noticeable), and now the feature seems to cost ~70MB on the linux kernel. It's probably acceptable.

**Comment by Byron** [comment]
I really can't wait for this commit to be merged, as I think I have a neat follow-up in the works: [This PR](https://github.com/Byron/gitoxide/pull/887) adds support for commit-graphs during iteration, and if done right it's possible to multi-thread the author-map that currently relying on single-threaded iteration. As this workload should scale perfectly per core, I'd expect the linux kernel to finish in 2s or so :) (at least if someone ran `git commit-graph write --no-progress --reachable` which those with huge repos probably do).

**Pull request merged**